File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/English/TimexConvertEnglish.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Globalization;

namespace Microsoft.Recognizers.Text.DataTypes.TimexExpression
{
    internal static class TimexConvertEnglish
    {
        public static string ConvertTimexToString(TimexProperty timex)
        {
            if (timex.IsPresent)
            {
                return TimexConstantsEnglish.Now;
            }

            if (timex.IsDateTimeRange)
            {
                return ConvertDateTimeRange(timex);
            }

            if (timex.IsDateRange)
            {
                return ConvertDateRange(timex);
            }

            if (timex.IsDuration)
            {
                return ConvertDuration(timex);
            }

            if (timex.IsTimeRange)
            {
                return ConvertTimeRange(timex);
            }

            // TODO: where appropriate delegate most the formatting delegate to Date.toLocaleString(options)
            if (timex.IsDateTime)
            {
                return ConvertDateTime(timex);
            }

            if (timex.IsDate)
            {
                return ConvertDate(timex);
            }

            if (timex.IsTime)
            {
                return ConvertTime(timex);
            }

            return string.Empty;
        }

        public static string ConvertTimexSetToString(TimexSet timexSet)
        {
            var timex = timexSet.Timex;
            if (timex.Types.Contains(Constants.TimexTypes.Duration))
            {
                return $"{TimexConstantsEnglish.Every} {ConvertTimexDurationToString(timex, false)}";
            }
            else
            {
                return $"{TimexConstantsEnglish.Every} {ConvertTimexToString(timex)}";
            }
        }

        public static string ConvertTime(TimexProperty timex)
        {
            if (timex.Hour == 0 && timex.Minute == 0 && timex.Second == 0)
            {
                return TimexConstantsEnglish.Midnight;
            }

            if (timex.Hour == 12 && timex.Minute == 0 && timex.Second == 0)
            {
                return TimexConstantsEnglish.Midday;
            }

            var hour = (timex.Hour == 0) ? "12" : (timex.Hour > 12) ? (timex.Hour - 12).Value.ToString(CultureInfo.InvariantCulture) : timex.Hour.Value.ToString(CultureInfo.InvariantCulture);
            var minute = (timex.Minute == 0 && timex.Second == 0) ? string.Empty : Constants.TimeTimexConnector + timex.Minute.Value.ToString(CultureInfo.InvariantCulture).PadLeft(2, '0');
            var second = (timex.Second == 0) ? string.Empty : Constants.TimeTimexConnector + timex.Second.Value.ToString(CultureInfo.InvariantCulture).PadLeft(2, '0');
            var period = timex.Hour < 12 ? Constants.AM : Constants.PM;

            return $"{hour}{minute}{second}{period}";
        }

        public static string ConvertDate(TimexProperty timex)
        {
            if (timex.DayOfWeek != null)
            {
                return TimexConstantsEnglish.Days[timex.DayOfWeek.Value - 1];
            }

            var parts = new List<string>(3);

            if (timex.DayOfMonth.HasValue)
            {
                var dayOfMonth = timex.DayOfMonth.Value;

                // Ordinals 11 to 13 are special in english as they end in th
                var abbreviation = TimexConstantsEnglish.DateAbbreviation[(dayOfMonth is > 9 and < 14 ? 9 : dayOfMonth) % 10];
                parts.Add($"{dayOfMonth}{abbreviation}");
            }

            if (timex.Month != null)
            {
                parts.Add(TimexConstantsEnglish.Months[timex.Month.Value - 1]);
            }

            if (timex.Year != null)
            {
                parts.Add(timex.Year.ToString());
            }

            return parts.Count > 0 ? string.Join(" ", parts) : string.Empty;
        }

        private static string ConvertDurationPropertyToString(decimal value, string property, bool includeSingleCount)
        {
            if (value == 1)
            {
                return includeSingleCount ? "1 " + property : property;
            }
            else
            {
                return $"{value} {property}{Constants.TimeDurationUnit}";
            }
        }

        private static string ConvertTimexDurationToString(TimexProperty timex, bool includeSingleCount)
        {
            var parts = new List<string>();
            if (timex.Years != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Years.Value, Constants.YearUnit, includeSingleCount));
            }

            if (timex.Months != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Months.Value, Constants.MonthUnit, includeSingleCount));
            }

            if (timex.Weeks != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Weeks.Value, Constants.WeekUnit, includeSingleCount));
            }

            if (timex.Days != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Days.Value, Constants.DayUnit, includeSingleCount));
            }

            if (timex.Hours != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Hours.Value, Constants.HourUnit, includeSingleCount));
            }

            if (timex.Minutes != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Minutes.Value, Constants.MinuteUnit, includeSingleCount));
            }

            if (timex.Seconds != null)
            {
                parts.Add(ConvertDurationPropertyToString(timex.Seconds.Value, Constants.SecondUnit, includeSingleCount));
            }

            return parts.Count > 0 ? string.Join(" ", parts) : string.Empty;
        }

        private static string ConvertDuration(TimexProperty timex)
        {
            return ConvertTimexDurationToString(timex, true);
        }

        private static string ConvertDateRange(TimexProperty timex)
        {
            if (timex.IsDate && !timex.IsDuration)
            {
                return ConvertDate(timex);
            }

            var parts = new List<string>();
            var start = timex;
            var end = TimexProperty.Empty;

            if (timex.Types.Count > 1)
            {
                var dateRange = TimexHelpers.ExpandDateTimeRange(timex);
                start = dateRange.Start;
                end = dateRange.End;
            }

            // output the start of the range
            ConvertRangePoint(parts, start);

            if (end == TimexProperty.Empty)
            {
                return parts.Count > 0 ? string.Join(" ", parts) : string.Empty;
            }

            // seasons are a little special, easier to handle them separately
            if (end.IsSeason)
            {
                if (start.Season != end.Season || start.Year != end.Year)
                {
                    parts.Add(TimexConstantsEnglish.To);
                    parts.Add(TimexConstantsEnglish.Seasons[end.Season]);
                    if (end.Year != null)
                    {
                        parts.Add(end.Year.Value.ToString(CultureInfo.InvariantCulture));
                    }
                }
                else
                {
                    TryAddDuration(parts, timex);
                }

                return string.Join(" ", parts);
            }

            parts.Add(TimexConstantsEnglish.To);

            // output the end of the range
            ConvertRangePoint(parts, end);

            return parts.Count > 0 ? string.Join(" ", parts) : string.Empty;
        }

        private static void ConvertRangePoint(List<string> parts, TimexProperty timex)
        {
            // Timex with season can only have an optional year
            if (timex.IsSeason)
            {
                parts.Add(TimexConstantsEnglish.Seasons[timex.Season]);
            }

            if (timex.Month != null)
            {
                var timexMonth = $"{TimexConstantsEnglish.Months[timex.Month.Value - 1]}";

                // first/second/third/fourth week of {month}
                if (timex.WeekOfMonth != null)
                {
                    parts.Add(TimexConstantsEnglish.Weeks[timex.WeekOfMonth.Value - 1]);
                    parts.Add(TimexConstantsEnglish.Week);
                    parts.Add(TimexConstantsEnglish.Of);
                }

                parts.Add(timexMonth);
            }

            if (timex.DayOfWeek != null)
            {
                parts.Add(TimexConstantsEnglish.Days[timex.DayOfWeek.Value - 1]);
            }

            if (timex.Weekend != null)
            {
                parts.Add(TimexConstantsEnglish.Weekend);
                parts.Add(TimexConstantsEnglish.Of);
            }

            if (timex.WeekOfYear != null)
            {
                parts.Add(TimexConstantsEnglish.Week);
                parts.Add(timex.WeekOfYear.Value.ToString(CultureInfo.InvariantCulture));
            }

            if (timex.Year != null)
            {
                parts.Add(timex.Year.Value.ToString(CultureInfo.InvariantCulture));
            }
        }

        private static string ConvertTimeRange(TimexProperty timex)
        {
            return TimexConstantsEnglish.DayParts[timex.PartOfDay];
        }

        private static string ConvertDateTime(TimexProperty timex)
        {
            return $"{ConvertTime(timex)} {ConvertDate(timex)}";
        }

        private static string ConvertDateTimeRange(TimexProperty timex)
        {
            var parts = new List<string>();

            bool splitDates = timex.Types.Contains(Constants.TimexTypes.DateRange);

            var range = splitDates
                ? TimexHelpers.ExpandDateTimeRange(timex)
                : TimexHelpers.ExpandTimeRange(timex);

            var start = range.Start;
            var end = range.End;

            splitDates &= start.DateFromTimex() != end.DateFromTimex();

            bool splitTimes = timex.IsTimeRange && !timex.IsPartOfDay;

            TryAddDate(parts, splitDates ? start : timex);
            TryAddTime(parts, splitTimes ? start : timex);

            if (splitDates)
            {
                parts.Add(TimexConstantsEnglish.To);
                TryAddDate(parts, end);
                TryAddTime(parts, end);
                TryAddDuration(parts, end);
            }
            else if (splitTimes)
            {
                parts.Add(TimexConstantsEnglish.To);
                TryAddTime(parts, end);
                TryAddDuration(parts, end);
            }

            if (parts.Count > 0)
            {
                return string.Join(" ", parts);
            }

            return string.Empty;
        }

        private static bool TryAddDate(List<string> list, TimexProperty timexProperty)
        {
            var ret = timexProperty.Types.Contains(Constants.TimexTypes.Date);
            if (ret)
            {
                list.Add(ConvertDate(timexProperty));
            }

            return ret;
        }

        private static bool TryAddTime(List<string> list, TimexProperty timexProperty)
        {
            var ret = timexProperty.Types.Contains(Constants.TimexTypes.Time) || timexProperty.Types.Contains(Constants.TimexTypes.TimeRange);
            if (ret)
            {
                list.Add(timexProperty.PartOfDay is not null ? ConvertTimeRange(timexProperty) : ConvertTime(timexProperty));
            }

            return ret;
        }

        private static bool TryAddDuration(List<string> list, TimexProperty timexProperty)
        {
            var ret = timexProperty.Types.Contains(Constants.TimexTypes.Duration);
            if (ret)
            {
                list.Add(ConvertDuration(timexProperty));
            }

            return ret;
        }

    }
}

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/english/TimexConvertEnglish.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.datatypes.timex.expression.english;

import com.microsoft.recognizers.datatypes.timex.expression.Constants;
import com.microsoft.recognizers.datatypes.timex.expression.TimexInference;
import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;
import com.microsoft.recognizers.datatypes.timex.expression.TimexSet;

import java.math.BigDecimal;
import java.util.HashSet;

public class TimexConvertEnglish {
    public static String convertTimexToString(TimexProperty timex) {
        HashSet<String> types = timex.getTypes().size() != 0 ? timex.getTypes() : TimexInference.infer(timex);

        if (types.contains(Constants.TimexTypes.PRESENT)) {
            return TimexConstantsEnglish.NOW;
        }

        if (types.contains(Constants.TimexTypes.DATE_TIME_RANGE)) {
            return TimexConvertEnglish.convertDateTimeRange(timex);
        }

        if (types.contains(Constants.TimexTypes.DATE_RANGE)) {
            return TimexConvertEnglish.convertDateRange(timex);
        }

        if (types.contains(Constants.TimexTypes.DURATION)) {
            return TimexConvertEnglish.convertDuration(timex);
        }

        if (types.contains(Constants.TimexTypes.TIME_RANGE)) {
            return TimexConvertEnglish.convertTimeRange(timex);
        }

        // TODO: where appropriate delegate most the formatting delegate to
        // Date.toLocaleString(options)
        if (types.contains(Constants.TimexTypes.DATE_TIME)) {
            return TimexConvertEnglish.convertDateTime(timex);
        }

        if (types.contains(Constants.TimexTypes.DATE)) {
            return TimexConvertEnglish.convertDate(timex);
        }

        if (types.contains(Constants.TimexTypes.TIME)) {
            return TimexConvertEnglish.convertTime(timex);
        }

        return new String();
    }

    public static String convertTimexSetToString(TimexSet timexSet) {
        TimexProperty timex = timexSet.getTimex();
        if (timex.getTypes().contains(Constants.TimexTypes.DURATION)) {
            return String.format("%1$s %2$s", TimexConstantsEnglish.EVERY,
                    TimexConvertEnglish.convertTimexDurationToString(timex, false));
        } else {
            return String.format("%1$s %2$s", TimexConstantsEnglish.EVERY,
                    TimexConvertEnglish.convertTimexToString(timex));
        }
    }

    public static String convertTime(TimexProperty timex) {
        if (timex.getHour() == 0 && timex.getMinute() == 0 && timex.getSecond() == 0) {
            return TimexConstantsEnglish.MIDNIGHT;
        }

        if (timex.getHour() == 12 && timex.getMinute() == 0 && timex.getSecond() == 0) {
            return TimexConstantsEnglish.MIDDAY;
        }

        String hour = (timex.getHour() == 0) ? "12"
                : (timex.getHour() > 12) ? String.valueOf(timex.getHour() - 12) : String.valueOf(timex.getHour());
        String minute = (timex.getMinute() == 0 && timex.getSecond() == 0) ? new String()
                : Constants.TIME_TIMEX_CONNECTOR
                        + String.format("%1$2s", String.valueOf(timex.getMinute())).replace(' ', '0');
        String second = (timex.getSecond() == 0) ? new String()
                : Constants.TIME_TIMEX_CONNECTOR
                        + String.format("%1$2s", String.valueOf(timex.getSecond())).replace(' ', '0');
        String period = timex.getHour() < 12 ? Constants.AM : Constants.PM;

        return String.format("%1$s%2$s%3$s%4$s", hour, minute, second, period);
    }

    public static String convertDate(TimexProperty timex) {
        if (timex.getDayOfWeek() != null) {
            return TimexConstantsEnglish.DAYS[timex.getDayOfWeek() - 1];
        }

        String date = String.valueOf(timex.getDayOfMonth());

        String abbreviation = TimexConstantsEnglish.DATE_ABBREVIATION[Integer
                .parseInt(String.valueOf(date.charAt(date.length() - 1)))];

        if (timex.getMonth() != null) {
            String month = TimexConstantsEnglish.MONTHS[timex.getMonth() - 1];
            if (timex.getYear() != null) {
                return String.format("%1$s%2$s %3$s %4$s", date, abbreviation, month, timex.getYear()).trim();
            }
            return String.format("%1$s%2$s %3$s", date, abbreviation, month);
        }
        return date.concat(abbreviation);
    }

    private static String convertDurationPropertyToString(BigDecimal value, String property,
            Boolean includeSingleCount) {
        if (value.intValue() == 1) {
            return includeSingleCount ? "1 " + property : property;
        } else {
            return String.format("%1$s %2$s%3$s", value, property, Constants.TIME_DURATION_UNIT);
        }
    }

    private static String convertTimexDurationToString(TimexProperty timex, Boolean includeSingleCount) {
        String result = new String();
        if (timex.getYears() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getYears(), Constants.YEAR_UNIT,
                    includeSingleCount);
        }

        if (timex.getMonths() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getMonths(), Constants.MONTH_UNIT,
                    includeSingleCount);
        }

        if (timex.getWeeks() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getWeeks(), Constants.WEEK_UNIT,
                    includeSingleCount);
        }

        if (timex.getDays() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getDays(), Constants.DAY_UNIT,
                    includeSingleCount);
        }

        if (timex.getHours() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getHours(), Constants.HOUR_UNIT,
                    includeSingleCount);
        }

        if (timex.getMinutes() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getMinutes(), Constants.MINUTE_UNIT,
                    includeSingleCount);
        }

        if (timex.getSeconds() != null) {
            result += TimexConvertEnglish.convertDurationPropertyToString(timex.getSeconds(), Constants.SECOND_UNIT,
                    includeSingleCount);
        }

        return result;
    }

    private static String convertDuration(TimexProperty timex) {
        return TimexConvertEnglish.convertTimexDurationToString(timex, true);
    }

    private static String convertDateRange(TimexProperty timex) {
        String season = (timex.getSeason() != null) ? TimexConstantsEnglish.SEASONS.get(timex.getSeason())
                : new String();

        String year = (timex.getYear() != null) ? timex.getYear().toString() : new String();

        if (timex.getWeekOfYear() != null) {
            if (timex.getWeekend() != null) {
                throw new UnsupportedOperationException();
            }
        }

        if (timex.getMonth() != null) {
            String month = TimexConstantsEnglish.MONTHS[timex.getMonth() - 1];
            if (timex.getWeekOfMonth() != null) {
                return String.format("%1$s week of %2$s", TimexConstantsEnglish.WEEKS[timex.getWeekOfMonth() - 1],
                        month);
            } else {
                return String.format("%1$s %2$s", month, year).trim();
            }
        }

        return String.format("%1$s %2$s", season, year).trim();
    }

    private static String convertTimeRange(TimexProperty timex) {
        return TimexConstantsEnglish.DAY_PARTS.get(timex.getPartOfDay());
    }

    private static String convertDateTime(TimexProperty timex) {
        return String.format("%1$s %2$s", TimexConvertEnglish.convertTime(timex),
                TimexConvertEnglish.convertDate(timex));
    }

    private static String convertDateTimeRange(TimexProperty timex) {
        if (timex.getTypes().contains(Constants.TimexTypes.TIME_RANGE)) {
            return String.format("%1$s %2$s", TimexConvertEnglish.convertDate(timex),
                    TimexConvertEnglish.convertTimeRange(timex));
        }

        // date + time + duration
        // - OR -
        // date + duration
        return new String();
    }
}
